File: packages/react-native-audio-api/common/cpp/system/AudioControls.cpp

#include "AudioControls.h"
#include "NowPlayingInfo.h"
#include "AudioSessionOptions.h"

#ifdef ANDROID
#include "AndroidAudioControls.h"
#else
#include "IOSAudioControls.h"
#endif

namespace audioapi {

AudioControls::AudioControls() {
#ifdef ANDROID
  audioControls_ = std::make_shared<AndroidAudioControls>();
#else
  audioControls_ = std::make_shared<IOSAudioControls>();
#endif
}

AudioControls::~AudioControls() {
  // TODO: cleanup everything if disable was not called
}

void AudioControls::init(std::shared_ptr<AudioSessionOptions> options) {
  audioControls_->init(options);
}

void AudioControls::updateOptions(std::shared_ptr<AudioSessionOptions> options) {
  audioControls_->updateOptions(options);
}

void AudioControls::disable() {
  audioControls_->disable();
}

void AudioControls::showNowPlayingInfo(std::shared_ptr<NowPlayingInfo> nowPlayingInfo) {
  audioControls_->showNowPlayingInfo(nowPlayingInfo);
}

void AudioControls::updateNowPlayingInfo(std::shared_ptr<NowPlayingInfo> nowPlayingInfo) {
  audioControls_->updateNowPlayingInfo(nowPlayingInfo);
}

void AudioControls::hideNowPlayingInfo() {
  audioControls_->hideNowPlayingInfo();
}

void AudioControls::addEventListener() {
  audioControls_->addEventListener();
}

} // namespace audioapi

File: packages/react-native-audio-api/src/system/AudioControls.ts
import type { IAudioControls } from '../interfaces';
import {
  IOSMode,
  IOSCategory,
  EventListener,
  NowPlayingInfo,
  AudioEventName,
  InterruptionMode,
  IOSCategoryOption,
  AudioEventCallback,
  AudioSessionOptions,
  NowPlayingInfoUpdate,
} from '../types';

const defaultOptions: AudioSessionOptions = {
  iosCategory: IOSCategory.Playback,
  iosMode: IOSMode.Default,
  iosCategoryOptions: [
    IOSCategoryOption.DuckOthers,
    IOSCategoryOption.MixWithOthers,
  ],
  interruptionMode: InterruptionMode.Automatic,
  androidForegroundService: true,
};

function withDefaultOptions(
  options?: AudioSessionOptions
): AudioSessionOptions {
  return {
    ...defaultOptions,
    ...(options ?? {}),
  };
}

class AudioControls {
  protected module(): IAudioControls {
    return global.__AudioAPIInstaller.getAudioControls();
  }

  init(options?: AudioSessionOptions) {
    this.module().init(withDefaultOptions(options));
  }

  updateOptions(options?: AudioSessionOptions) {
    this.module().updateOptions(withDefaultOptions(options));
  }

  disable() {
    this.module().disable();
  }

  showNowPlayingInfo(info: NowPlayingInfo) {
    this.module().showNowPlayingInfo(info);
  }

  updateNowPlayingInfo(info: NowPlayingInfoUpdate) {
    this.module().updateNowPlayingInfo(info);
  }

  hideNowPlayingInfo() {
    this.module().hideNowPlayingInfo();
  }

  addEventListener<EN extends AudioEventName>(
    eventName: EN,
    eventHandler: AudioEventCallback<EN>
  ): EventListener {
    this.module().addEventListener(eventName, eventHandler);

    return {
      // TODO:
      remove: () => {},
    };
  }
}

export default new AudioControls();
